src/vec_simd.rs

```diff
@@ -67,6 +81,100 @@ pub fn scaled_add(mut u: ArrayViewMut1<f32>, v: ArrayView1<f32>, a: f32) {
     )
 }
 
+#[cfg(all(target_arch = "aarch64", feature = "neon"))]
+pub mod neon {
+    use std::arch::aarch64::*;
+
+    use ndarray::{ArrayView1, ArrayViewMut1};
+
+    use crate::vec_simd::{dot_unvectorized, scale_unvectorized, scaled_add_unvectorized};
+
+    #[inline(always)]
+    #[allow(dead_code)]
+    pub(crate) unsafe fn neon_add(values: float32x4_t) -> f32 {
+        // vaddvq_f32 is missing in std::arch::aarch64?
+        let sums = vadd_f32(vget_high_f32(values), vget_low_f32(values));
+        vget_lane_f32::<0>(sums) + vget_lane_f32::<1>(sums)
+    }
+
+    #[target_feature(enable = "neon")]
+    pub unsafe fn dot(u: ArrayView1<f32>, v: ArrayView1<f32>) -> f32 {
+        assert_eq!(u.len(), v.len());
+
+        let mut u = u
+            .as_slice()
+            .expect("Cannot apply SIMD instructions on non-contiguous data.");
+        let mut v = &v
+            .as_slice()
+            .expect("Cannot apply SIMD instructions on non-contiguous data.")[..u.len()];
+
+        let mut sums = vdupq_n_f32(0f32);
+        while u.len() >= 4 {
```

Comment:
I like the `chunks_exact()` / `chunks()` API on slices to iterate over chunks, if you prefer to stick with the manual stepping & slicing, that's fine with me, too!

Reply:
Cool `chunks_exact`, is really a nice addition to the slice API, I missed that! I think this would be nicer than the loops as they are now, but I'll have to check that bounds checks are still correctly eliminated. So, I'll make a separate PR out of that if it works.

Reply:
Sounds good!

Reply:
I have only inspected the machine code for ARM Neon so far, but it does seem to eliminate unnecessary bounds checking. The code machine code sometimes ends being slightly different, but with no practical effect. Wanted to show the assembly for `dot` though, compilers can be so clever:

<img width="695" alt="Screen Shot 2021-11-22 at 20 22 41" src="https://user-images.githubusercontent.com/49398/142922451-7a23dc1f-42c8-4a9a-8311-f0e7bd045020.png">

```
and                 x11, x0, #0xfffffffffffffffc
```

`x11` acts as the counter in the main loop. First, it is masked, so that its value will always be a multiple of 4. So, e.g. if the dimensionality is 302, it will become 300.

```
neg                 x11, x11
```

Negates the counter, so it becomes `-300`. Later in the loop:

```
adds                x11, x11, #4
b.ne                "finalfrontier::vec_simd::neon::dot::hcc5a53bb5a65e6ba+0xa4"
```

So, rather than counting from 0 to 300, the loop counts from -300 to 0. The `adds` instruction adds 4 to the register, in contrast to `add` it sets some flags. One of the flags that it sets is `Z`, which is `1` is the addition resulted in `0` or `0` otherwise.  `b.ne` jumps to the given label if `Z == 0`.

So, it's counting upwards from -300, because then hitting zero sets the stopping condition. If it would count from 0 to 300 an additional instruction would be needed to compare the register against 300. I guess another possibility would be to count down from 300.

Anyway, this is all very unimportant, but I thought the masking + negation was clever :).